backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java

```diff
@@ -237,6 +236,104 @@ public void revokePermission(UUID bucketId, UUID userId) {
     permissionRepository.delete(permission);
   }
 
+  @Override
+  public PermissionResponse grantPermissionForFile(PermissionGrantRequest permissionGrantRequest) {
+    UserObjectAccessEntity permission = new UserObjectAccessEntity();
+
+    UserEntity currentUser =
+        userRepository
+            .findById(currentUserService.getLoggedUser().getUuid())
+            .orElseThrow(UserNotFoundException::new);
+
+    UserEntity targetUser =
+        userRepository
+            .findById(permissionGrantRequest.getGrantedTo())
+            .orElseThrow(UserNotFoundException::new);
+
+    AccessTypeEntity accessType =
+        accessTypeRepository
+            .findByName(permissionGrantRequest.getAccess())
+            .orElseThrow(AccessTypeNotFoundException::new);
+
+    FileMetadataEntity file =
+        fileMetadataRepository
+            .findById(permissionGrantRequest.getObjectId())
+            .orElseThrow(BucketNotFoundException::new);
+
+    if (file.isDeleted()) {
+      throw new FileAlreadyDeletedException();
+    }
+
+    permission.setGrantedBy(currentUser);
+    permission.setGrantedTo(targetUser);
+    permission.setAccess(accessType);
+    permission.setObject(file);
+    permission.setId(
+        new UserObjectAccessKey(permission.getGrantedTo().getId(), permission.getObject().getId()));
+
+    currentUserHasPermissionForBucket(permissionGrantRequest.getObjectId(), AccessType.OWNER);
+
+    if (permissionRepository.existsById(
+        new UserObjectAccessKey(targetUser.getId(), file.getId()))) {
+      throw new PermissionAlreadyExistsException();
+    }
+
+    UserObjectAccessEntity savedPermission = permissionRepository.save(permission);
+
+    grantViewPermissionToAllFoldersUpstream(file.getParentFolder(), currentUser, targetUser);
+
+    permissionGrantRequest.setObjectId(file.getBucket().getId());
+    permissionGrantRequest.setAccess("view");
+    grantPermissionForBucket(permissionGrantRequest);
+
+    log.info(
+        "User {} ({}) granted {} permission for file {} ({}) to user {} ({})",
+        permission.getGrantedBy().getEmail(),
+        permission.getGrantedBy().getId(),
+        permission.getAccess().getName(),
+        permission.getObject().getName(),
+        permission.getObject().getId(),
+        permission.getGrantedTo().getEmail(),
+        permission.getGrantedTo().getId());
+
+    return modelMapper.map(savedPermission, PermissionResponse.class);
+  }
+
+  private void grantViewPermissionToAllFoldersUpstream(
```

Comment:
right now, this method  is basically grant permission for folder

Reply:
Forgot to write the call to itself :D